lib/redis/connection/ruby.rb

```diff
@@ -34,20 +35,47 @@ def write_timeout=(timeout)
       end
 
       def read(nbytes)
-        result = @buffer.slice!(0, nbytes)
+        result = @buffer.byteslice(@offset, nbytes)
+        @offset += result.bytesize
 
         buffer = String.new(capacity: nbytes, encoding: Encoding::ASCII_8BIT)
         result << _read_from_socket(nbytes - result.bytesize, buffer) while result.bytesize < nbytes
 
         result
       end
 
+      def getbyte
+        fill_buffer if empty_buffer?
+
+        @buffer[@offset].tap do |r|
+          @offset += 1
+        end
+      end
+
       def gets
-        while (crlf = @buffer.index(CRLF)).nil?
-          @buffer << _read_from_socket(16_384)
+        while (crlf = @buffer.index(CRLF, @offset)).nil?
+          fill_buffer
+        end
+
+        end_index = crlf + CRLF.bytesize - 1
+        @buffer.byteslice(@offset..end_index).tap do
+          @offset = end_index + 1
         end
+      end
+
+      def fill_buffer
+        data = _read_from_socket(16_384)
```

Comment:
Related to the other comment, I noticed that we don't pre-allocate the buffer in `#gets`. Shouldn't they be consistent?

I just left the code as-is though since this is not related to the current change.